### ISSUE TITLE
Cannot execute the persistent action of helm-imenu when helm-buffer overrides helm-current-buffer

### ISSUE THREAD

**Issue by xuchunyang** [emacs-helm/helm#1036]
The steps to reproduce:

1. Make two windows side by side, the right is init.el
2. Switch to the right window
3. M-x helm-imenu (`*helm imenu*` buffer now overrides `init.el` buffer)
4. C-j ("helm-execute-persistent-action") doesn't work, it says: 
```
 apply: Marker points into wrong buffer: #<marker at 1543 in init.el>
```

The following change can fix it but I'm not sure if it is a good solution.
```diff
diff --git a/helm-imenu.el b/helm-imenu.el
index 2371068..28593b5 100644
--- a/helm-imenu.el
+++ b/helm-imenu.el
@@ -127,6 +127,7 @@

 (defun helm-imenu-persistent-action (candidate)
   "Default persistent action for `helm-source-imenu'."
+  (switch-to-buffer helm-current-buffer)
   (imenu candidate)
   (helm-highlight-current-line))
```

**Comment by thierryvolpiatto** [comment]
Chunyang Xu <notifications@github.com> writes:

> The steps to reproduce:
>
>  1. Make two windows side by side, the right is init.el
>  2. Switch to the right window
>  3. M-x helm-imenu (*helm imenu* buffer now overrides init.el buffer)
>  4. C-j ("helm-execute-persistent-action") doesn't work, it says:
>
>  apply: Marker points into wrong buffer: #<marker at 1543 in init.el>

I can't reproduce, did you try from ./emacs-helm.sh ?

> The following change can fix it but I'm not sure if it is a good solution.

It is not, this should be handled at lower level.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by xuchunyang** [comment]
> I can't reproduce, did you try from ./emacs-helm.sh ?

Just tried, same issue:

```
Debugger entered--Lisp error: (error "Marker points into wrong buffer" #<marker at 905 in helm-imenu.el>)
  imenu-default-goto-function(#("helm-imenu" 0 10 (helm-imenu-type "Types")) #<marker at 905 in helm-imenu.el>)
  apply(imenu-default-goto-function #("helm-imenu" 0 10 (helm-imenu-type "Types")) #<marker at 905 in helm-imenu.el> nil)
  imenu((#("helm-imenu" 0 10 (helm-imenu-type "Types")) . #<marker at 905 in helm-imenu.el>))
  helm-imenu-persistent-action((#("helm-imenu" 0 10 (helm-imenu-type "Types")) . #<marker at 905 in helm-imenu.el>))
  apply(helm-imenu-persistent-action (#("helm-imenu" 0 10 (helm-imenu-type "Types")) . #<marker at 905 in helm-imenu.el>))
  (setq result (apply fn args))
  (while (consp --cl-var--) (setq fn (car --cl-var--)) (setq result (apply fn args)) (setq --cl-var-- (cdr --cl-var--)))
  (let* ((result nil) (--cl-var-- funs) (fn nil)) (while (consp --cl-var--) (setq fn (car --cl-var--)) (setq result (apply fn args)) (setq --cl-var-- (cdr --cl-var--))) result)
  (let ((helm-source-name (assoc-default (quote name) source)) (helm-current-source source) (funs (if (functionp functions) (list functions) functions))) (helm-log "helm-source-$
  helm-funcall-with-source(((name . "Imenu") (candidates . helm-imenu-candidates) (keymap keymap (right . helm-execute-persistent-action) (left . helm-maybe-exit-minibuffer) (M-$
  (progn (helm-funcall-with-source source action (helm-coerce-selection selection source)))
  (if (and selection action) (progn (helm-funcall-with-source source action (helm-coerce-selection selection source))))
  (let ((source (or helm-saved-current-source (helm-get-current-source))) non-essential) (setq selection (or selection (helm-get-selection) (and (assoc (quote accept-empty) sour$
  helm-execute-selection-action-1(nil helm-imenu-persistent-action t)
  (let ((display-buffer-function (quote helm-persistent-action-display-buffer))) (helm-execute-selection-action-1 nil (or fn (helm-get-actions-from-current-source)) t) (helm-log$
  (let ((helm-in-persistent-action t)) (let ((display-buffer-function (quote helm-persistent-action-display-buffer))) (helm-execute-selection-action-1 nil (or fn (helm-get-actio$
  (progn (if no-split (helm-select-persistent-action-window) (helm-select-persistent-action-window (or split-onewindow helm-onewindow-p))) (helm-log "current-buffer = %S" (curre$
  (unwind-protect (progn (if no-split (helm-select-persistent-action-window) (helm-select-persistent-action-window (or split-onewindow helm-onewindow-p))) (helm-log "current-buf$
  (save-current-buffer (unwind-protect (progn (if no-split (helm-select-persistent-action-window) (helm-select-persistent-action-window (or split-onewindow helm-onewindow-p))) ($
  (let ((save-selected-window--state (internal--before-save-selected-window))) (save-current-buffer (unwind-protect (progn (if no-split (helm-select-persistent-action-window) (h$
  (progn (select-window (car save-selected-window--state) (quote norecord)) (let ((save-selected-window--state (internal--before-save-selected-window))) (save-current-buffer (un$
  (unwind-protect (progn (select-window (car save-selected-window--state) (quote norecord)) (let ((save-selected-window--state (internal--before-save-selected-window))) (save-cu$
  (save-current-buffer (unwind-protect (progn (select-window (car save-selected-window--state) (quote norecord)) (let ((save-selected-window--state (internal--before-save-select$
  (let ((save-selected-window--state (internal--before-with-selected-window (helm-window)))) (save-current-buffer (unwind-protect (progn (select-window (car save-selected-window$
  (let* ((attr-val (assoc-default attr (helm-get-current-source))) (fn (if (and (consp attr-val) (not (functionp attr-val))) (car attr-val) attr-val)) (no-split (and (consp attr$
  (progn (helm-log "executing persistent-action") (let* ((attr-val (assoc-default attr (helm-get-current-source))) (fn (if (and (consp attr-val) (not (functionp attr-val))) (car$
  (let* ((attr (if --cl-rest-- (car-safe (prog1 --cl-rest-- (setq --cl-rest-- (cdr --cl-rest--)))) (quote persistent-action))) (split-onewindow (car-safe (prog1 --cl-rest-- (set$
  helm-execute-persistent-action()
  call-interactively(helm-execute-persistent-action nil nil)
  command-execute(helm-execute-persistent-action)
  read-from-minibuffer("pattern: " nil (keymap (f13 lambda nil (interactive) (helm-select-nth-action 12)) (f12 lambda nil (interactive) (helm-select-nth-action 11)) (f11 lambda $
  (progn (add-hook (quote minibuffer-setup-hook) setup-hook) (read-from-minibuffer (or any-prompt "pattern: ") any-input helm-map nil hist tap helm-inherit-input-method))
  (unwind-protect (progn (add-hook (quote minibuffer-setup-hook) setup-hook) (read-from-minibuffer (or any-prompt "pattern: ") any-input helm-map nil hist tap helm-inherit-input$
  (let (setup-hook) (setq setup-hook (function (lambda nil (remove-hook (quote minibuffer-setup-hook) setup-hook) (funcall (function (lambda nil (helm--minor-mode 1) (setq minor$
  (unwind-protect (let (setup-hook) (setq setup-hook (function (lambda nil (remove-hook (quote minibuffer-setup-hook) setup-hook) (funcall (function (lambda nil ... ... ...)))))$
  (let ((tap (or any-default (save-current-buffer (set-buffer (or (and ... helm-current-buffer) (setq helm-current-buffer ...))) (thing-at-point (quote symbol)))))) (unwind-prot$
  (cond ((and helm-execute-action-at-once-if-one (not source-delayed-p) (= (helm-get-candidate-number) 1)) (ignore)) ((and helm-quit-if-no-candidate (not source-delayed-p) (= (h$
  (let* ((src (helm-get-current-source)) (src-keymap (assoc-default (quote keymap) src)) (hist (or (and any-history (symbolp any-history) any-history) (assoc-default (quote hist$
  (save-current-buffer (set-buffer (helm-buffer-get)) (let* ((src (helm-get-current-source)) (src-keymap (assoc-default (quote keymap) src)) (hist (or (an
```

**Comment by michael-heerdegen** [comment]
I cannot reproduce, too.  Could it be related to something in your config?

**Comment by xuchunyang** [comment]
> I cannot reproduce, too. Could it be related to something in your config?

No, because I've already test under ./emacs-helm.sh. It's pretty weird. I guess I didn't explain myself clear due my quite limited English knowledge.

As my patch says, we might need to switch to `helm-current-buffer` first and call `(imenu candidate)`, just like [helm-occur's helm-moccur-persistent-action](https://github.com/emacs-helm/helm/blob/c63cbe9c11caffbb5b2eca9a3c06f6eb658fcf08/helm-regexp.el#L245)

**Comment by thierryvolpiatto** [comment]
Chunyang Xu <notifications@github.com> writes:

>     I can't reproduce, did you try from ./emacs-helm.sh ?
>    
> Just tried, same issue:

Ok, I could reproduce, this happen when there is two windows displaying
two different buffers, will look into this as soon as possible.

Thanks.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by thierryvolpiatto** [comment]
Should work now in exp branch.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by xuchunyang** [comment]
Yes, it is. I've just tested. Thanks for the fix.

The default (without enableing `helm-autoresize-mode`) way of using window by helm has also been changed a bit: when invoking helm at the rightmost window, helm now maximize helm-current-buffer and popup helm-buffer from the bottom instead of displaying helm-buffer in helm-current-buffer. I think this is what you want.  I just want to say user who doesn't enabling `helm-autoresize-mode` will notice this change.
